## 38. 프로젝트 루트 확인

In [72]:
from pathlib import Path

project_root = Path.cwd()

print("현재 project 루트:", project_root)

if project_root.name == "notebooks":
    project_root = project_root.parent
    
data_dir = project_root / "data" / "raw"

print("프로젝트 루트:", project_root)
print("데이터 폴더:", data_dir)
print("데이터 폴더 존재:", data_dir.exists())

현재 project 루트: c:\dev\ai-data-analysis-repository\notebooks
프로젝트 루트: c:\dev\ai-data-analysis-repository
데이터 폴더: c:\dev\ai-data-analysis-repository\data\raw
데이터 폴더 존재: True


## 39. 필수 파일 확인

In [73]:
required_files = [
    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",
]

for file_name in required_files:
    file_path = data_dir / file_name
    print(file_name, file_path.exists(), file_path.stat().st_size if file_path.exists() else 0)

customers.csv True 5641
products.csv True 4101
orders.csv True 11696
order_items.csv True 15355


## 40. 데이터 불러오기

In [74]:
import pandas as pd
customers = pd.read_csv(data_dir / "customers.csv")
products = pd.read_csv(data_dir / "products.csv")
orders = pd.read_csv(data_dir / "orders.csv")
order_items = pd.read_csv(data_dir / "order_items.csv")

## 41. DataFrame 사전 만들기

In [75]:

datasets = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}

## 42. 크기 요약

In [76]:
shape_summary = pd.DataFrame(

    [

        {
            "dataset": name,
            "rows": df.shape[0],
            "columns": df.shape[1],
        }

        for name, df in datasets.items()

    ]

)


print(shape_summary)

       dataset  rows  columns
0    customers   150        6
1     products   100        4
2       orders   301        5
3  order_items   765        5


## 43. 한 행의 의미 기록

## 데이터셋별 한 행의 의미

- customers: 고객 한 명
- products: 상품 한 개
- orders: 주문 한 건
- order_items: 주문에 포함된 상품 한 항목

 ## 44. 처음과 마지막 행 확인

In [77]:
for name, df in datasets.items():
    print(f"\n[{name}] 앞부분")
    display(df.head())
    print(f"[{name}] 뒷부분")
    display(df.tail())


[customers] 앞부분


,customer_id,name,gender,age,city,signup_date
0,1,김수민,F,19,광주,2024-07-17
1,2,김정호,F,32,대구,2025-11-30
2,3,이경수,F,61,성남,2024-07-10
3,4,조영호,F,55,울산,2026-05-11
4,5,이예원,F,19,부산,2024-10-11


[customers] 뒷부분


,customer_id,name,gender,age,city,signup_date
145,146,김숙자,M,61,성남,2026-01-20
146,147,이정남,M,19,부산,2025-03-11
147,148,오도현,M,29,고양,2026-07-13
148,149,김정자,M,20,부산,2024-11-16
149,150,조미영,M,40,대전,2026-01-01



[products] 앞부분


,product_id,product_name,category,price
0,1,전자기기 상품 001,전자기기,160000
1,2,도서 상품 002,도서,34000
2,3,전자기기 상품 003,전자기기,152000
3,4,생활용품 상품 004,생활용품,70000
4,5,식품 상품 005,식품,186000


[products] 뒷부분


,product_id,product_name,category,price
95,96,생활용품 상품 096,생활용품,112000
96,97,전자기기 상품 097,전자기기,154000
97,98,스포츠 상품 098,스포츠,10000
98,99,뷰티 상품 099,뷰티,200000
99,100,도서 상품 100,도서,102000



[orders] 앞부분


,order_id,customer_id,order_date,payment_method,order_status
0,1,123,2026-06-04,card,completed
1,2,77,2025-08-20,naver_pay,cancelled
2,3,138,2025-12-17,bank_transfer,cancelled
3,4,57,2026-02-27,kakao_pay,cancelled
4,5,125,2026-01-18,card,cancelled


[orders] 뒷부분


,order_id,customer_id,order_date,payment_method,order_status
296,297,22,2026-02-12,bank_transfer,completed
297,298,64,2026-01-22,naver_pay,cancelled
298,299,135,2025-09-05,bank_transfer,completed
299,300,96,2026-03-25,kakao_pay,cancelled
300,301,300,2026-09-08-06,kakao_pay,cancelled



[order_items] 앞부분


,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


[order_items] 뒷부분


,order_item_id,order_id,product_id,quantity,unit_price
760,761,298,40,4,174000
761,762,299,8,2,189000
762,763,299,12,4,175000
763,764,300,59,5,32000
764,765,499,200,3,128000


## 45. 컬럼명 확인

In [78]:

for name, df in datasets.items():
    print(name, df.columns.tolist())

customers ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
products ['product_id', 'product_name', 'category', 'price']
orders ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
order_items ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']


## 46. 컬럼명 공백 점검

In [79]:
for name, df in datasets.items():

    suspicious = [
        col for col in df.columns
        if col != col.strip()
    ]

    print(name, suspicious)

customers []
products []
orders []
order_items []


## 47. info() 실행

In [80]:
for name, df in datasets.items():
    print(f"\n===== {name} =====")
    df.info()



===== customers =====
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB

===== products =====
<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   product_id    100 non-null    int64
 1   product_name  100 non-null    str  
 2   category      100 non-null    str  
 3   price         100 non-null    int64
dtypes: int64(2), str(2)
memory usage: 3.3 KB

===== orders =====
<class 'pandas.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 5 columns):

## 48. dtype 요약표

In [81]:
dtype_rows = []

for name, df in datasets.items():
    for column, dtype in df.dtypes.items():
        dtype_rows.append(
            {
                "dataset": name,
                "column": column,
                "dtype": str(dtype),
            }
        )
dtype_summary = pd.DataFrame(dtype_rows)
dtype_summary

,dataset,column,dtype
0,customers,customer_id,int64
1,customers,name,str
2,customers,gender,str
3,customers,age,int64
4,customers,city,str
5,customers,signup_date,str
6,products,product_id,int64
7,products,product_name,str
8,products,category,str
9,products,price,int64


## 49. 날짜형 후보 확인

In [82]:

date_columns = {
    "customers": ["signup_date"],
    "orders": ["order_date"],
}

for dataset_name, columns in date_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        converted = pd.to_datetime(
            df[column],
            errors="coerce"
        )

        print(
            dataset_name,
            column,
            "원본 dtype:", df[column].dtype,
            "변환 실패:", converted.isna().sum()
        )

customers signup_date 원본 dtype: str 변환 실패: 0
orders order_date 원본 dtype: str 변환 실패: 1


## 50. 결측치 개수

In [83]:
for name, df in datasets.items():
    print(f"\n[{name}]")
    display(df.isna().sum().to_frame("missing_count"))


[customers]


,missing_count
customer_id,0
name,0
gender,0
age,0
city,0
signup_date,0



[products]


,missing_count
product_id,0
product_name,0
category,0
price,0



[orders]


,missing_count
order_id,0
customer_id,0
order_date,0
payment_method,0
order_status,0



[order_items]


,missing_count
order_item_id,0
order_id,0
product_id,0
quantity,0
unit_price,0


## 51. 결측치 비율

In [84]:
missing_summaries = []

for name, df in datasets.items():
    summary = pd.DataFrame(
        {
            "missing_count": df.isna().sum(),
            "missing_rate_pct": (
                df.isna().mean() * 100
            ).round(2),
        }

    )

    summary.insert(0, "dataset", name)
    summary.insert(1, "column", summary.index)
    missing_summaries.append(summary.reset_index(drop=True))

missing_summary = pd.concat(
    missing_summaries,
    ignore_index=True
)

missing_summary

,dataset,column,missing_count,missing_rate_pct
0,customers,customer_id,0,0.0
1,customers,name,0,0.0
2,customers,gender,0,0.0
3,customers,age,0,0.0
4,customers,city,0,0.0
5,customers,signup_date,0,0.0
6,products,product_id,0,0.0
7,products,product_name,0,0.0
8,products,category,0,0.0
9,products,price,0,0.0


## 52. 전체 행 중복

In [85]:

for name, df in datasets.items():

    print(name, "전체 중복:", df.duplicated().sum())

customers 전체 중복: 0
products 전체 중복: 0
orders 전체 중복: 0
order_items 전체 중복: 0


## 53. 주요 ID 점검

In [86]:
primary_keys = {

    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",

}

id_check_rows = []
for name, key in primary_keys.items():
    df = datasets[name]
    id_check_rows.append(
        {
            "dataset": name,
            "key": key,
            "missing": df[key].isna().sum(),
            "duplicates": df[key].duplicated().sum(),
            "unique": df[key].nunique(),
            "rows": len(df),
        }
    )

id_check_summary = pd.DataFrame(id_check_rows)
id_check_summary

,dataset,key,missing,duplicates,unique,rows
0,customers,customer_id,0,0,150,150
1,products,product_id,0,0,100,100
2,orders,order_id,0,0,301,301
3,order_items,order_item_id,0,0,765,765


## 54. 발견 내용을 기록합니다

## 데이터 품질 발견 기록

| 데이터셋        | 컬럼            | 문제 종류     | 건수 | 해석                        | 처리 여부                      |
| ----------- | ------------- | --------- | -: | ------------------------- | -------------------------- |
| customers   | signup_date   | 날짜 자료형    |  0 | 날짜 변환 실패는 없지만 현재 문자열 자료형임, datetime 변환 필요 | N/A |
| orders      | order_date    | 날짜 자료형    |  0 | 날짜 변환 실패는 없지만 현재 문자열 자료형임, datetime 변환 필요 | N/A |


## 55. 숫자형 기본 통계

In [87]:
print(customers[["age"]].describe())
print(products[["price"]].describe())
print(order_items[["quantity", "unit_price"]].describe())

              age
count  150.000000
mean    42.086667
std     15.613166
min     19.000000
25%     29.000000
50%     40.000000
75%     57.000000
max     69.000000
               price
count     100.000000
mean   110040.000000
std     56433.910574
min      5000.000000
25%     65750.000000
50%    112000.000000
75%    161000.000000
max    200000.000000
         quantity     unit_price
count  765.000000     765.000000
mean     3.053595  108586.928105
std      1.409951   56963.792455
min      1.000000    5000.000000
25%      2.000000   62000.000000
50%      3.000000  111000.000000
75%      4.000000  161000.000000
max      5.000000  200000.000000


## 56. 범주형 값 확인

In [88]:
category_columns = {

    "customers": ["gender", "city"],
    "products": ["category"],
    "orders": ["payment_method", "order_status"],

}

for dataset_name, columns in category_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        print(f"\n[{dataset_name}.{column}]")
        display(
            df[column]
            .value_counts(dropna=False)
            .to_frame("count")
        )


[customers.gender]


,count
gender,
F,84
M,66



[customers.city]


,count
city,
성남,21
광주,17
부산,16
대구,15
서울,15
울산,14
인천,14
대전,14
수원,13



[products.category]


,count
category,
스포츠,19
전자기기,17
생활용품,16
뷰티,16
도서,14
패션,11
식품,7



[orders.payment_method]


,count
payment_method,
kakao_pay,80
naver_pay,77
bank_transfer,74
card,70



[orders.order_status]


,count
order_status,
completed,184
cancelled,65
refunded,52


## 57. 날짜 범위 확인

In [89]:

signup_dates = pd.to_datetime(

    customers["signup_date"],

    errors="coerce"

)

order_dates = pd.to_datetime(

    orders["order_date"],

    errors="coerce"

)

print("가입일 변환 실패:", signup_dates.isna().sum())

print("가입일 범위:", signup_dates.min(), "~", signup_dates.max())

print("주문일 변환 실패:", order_dates.isna().sum())

print("주문일 범위:", order_dates.min(), "~", order_dates.max())

가입일 변환 실패: 0
가입일 범위: 2023-08-11 00:00:00 ~ 2026-07-25 00:00:00
주문일 변환 실패: 1
주문일 범위: 2025-08-06 00:00:00 ~ 2026-08-05 00:00:00


## 58. 날짜 관계 점검 준비

주문일과 고객 가입일을 직접 비교하려면 두 데이터를 연결해야 합니다.

이번 장에서는 구조와 키가 연결 가능한지 확인하고, 실제 병합과 분석은 이후 장에서 본격적으로 진행합니다.


## 59. 참조하지 못한 고객 ID

In [90]:
unknown_customer_ids = sorted(
    set(orders["customer_id"].dropna())
    - set(customers["customer_id"].dropna())
)
print("존재하지 않는 customer_id:", unknown_customer_ids)

존재하지 않는 customer_id: [300]


## 60. 참조하지 못한 주문 ID

In [91]:
unknown_order_ids = sorted(
    set(order_items["order_id"].dropna())
    - set(orders["order_id"].dropna())
)
print("존재하지 않는 order_id:", unknown_order_ids)

존재하지 않는 order_id: [499]


## 61. 참조하지 못한 상품 ID

In [92]:
unknown_product_ids = sorted(
    set(order_items["product_id"].dropna())
    - set(products["product_id"].dropna())
)
print("존재하지 않는 product_id:", unknown_product_ids)

존재하지 않는 product_id: [200]


## 62. 반복 점검 함수

In [93]:
def inspect_dataframe(
    name: str,
    df: pd.DataFrame,
    primary_key: str | None = None,
) -> dict:
    result = {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
        "column_names": df.columns.tolist(),
        "dtypes": {column: str(dtype) for column, dtype in df.dtypes.items()},
        "missing_total": int(df.isna().sum().sum()),
        "duplicate_rows": int(df.duplicated().sum()),
    }
    if primary_key is not None:
        result["primary_key"] = primary_key
        result["key_missing"] = int(df[primary_key].isna().sum())
        result["key_duplicates"] = int(df[primary_key].duplicated().sum())
        result["key_unique"] = int(df[primary_key].nunique())
    return result

In [94]:
inspection_results = []
for name, df in datasets.items():
    inspection_results.append(
        inspect_dataframe(
            name,
            df,
            primary_keys[name],
        )
    )
inspection_results

[{'dataset': 'customers',
  'rows': 150,
  'columns': 6,
  'column_names': ['customer_id',
   'name',
   'gender',
   'age',
   'city',
   'signup_date'],
  'dtypes': {'customer_id': 'int64',
   'name': 'str',
   'gender': 'str',
   'age': 'int64',
   'city': 'str',
   'signup_date': 'str'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'customer_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 150},
 {'dataset': 'products',
  'rows': 100,
  'columns': 4,
  'column_names': ['product_id', 'product_name', 'category', 'price'],
  'dtypes': {'product_id': 'int64',
   'product_name': 'str',
   'category': 'str',
   'price': 'int64'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'product_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 100},
 {'dataset': 'orders',
  'rows': 301,
  'columns': 5,
  'column_names': ['order_id',
   'customer_id',
   'order_date',
   'payment_method',
   'order_status'],
  'dtypes': {'order_id': 'int6

## 63. LLM에 전달할 안전한 구조 요약

In [95]:
safe_schema_summary = {
    "datasets": inspection_results,
    "relationships": {
        "orders.customer_id": {
            "references": "customers.customer_id",
            "unknown_count": len(unknown_customer_ids),
        },
        "order_items.order_id": {
            "references": "orders.order_id",
            "unknown_count": len(unknown_order_ids),
        },
        "order_items.product_id": {
            "references": "products.product_id",
            "unknown_count": len(unknown_product_ids),
        },
    },
}
print(safe_schema_summary)

{'datasets': [{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 100, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 100}, {'dataset': 'orders', 'rows': 301, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status'

## 64. LLM 구조 점검 프롬프트


나는 온라인 쇼핑몰 데이터 분석을 준비하고 있습니다.

아래 정보는 원본 행이 아니라 구조 점검 요약입니다.

[여기에 safe_schema_summary의 내용을 붙여 넣기]

다음을 검토해 주세요.

1. 각 데이터셋에서 한 행이 무엇을 의미할 가능성이 높은지 설명해 주세요.

2. 주요 ID와 외래 키를 구분해 주세요.

3. 즉시 삭제하면 안 되는 결측치·중복 사례를 알려 주세요.

4. 추가로 점검할 구조 문제를 우선순위대로 제안해 주세요.

5. 실제 데이터에 없는 컬럼이나 관계를 가정하지 마세요.

6. 각 제안에 대해 확인할 pandas 코드의 목적만 설명해 주세요.

개인정보나 원본 고객 행은 제공하지 않았습니다.

## 65. LLM 답변 검증표

| 검증 항목 | 확인 내용 |
|---|---|
|컬럼 존재|실제 컬럼명과 일치하는가?|
|한 행의 의미|데이터셋 역할과 일치하는가?|
|ID 해석|식별자를 수치 변수로 해석하지 않았는가?|
|중복 해석|반복 가능한 외래 키를 오류로 판단하지 않았는가?|
|결측 해석|업무 의미 확인 없이 삭제를 권하지 않는가?|
|관계|실제 연결 키와 일치하는가?|
|개인정보|추가 원본 행 제공을 요구하지 않는가?|

# 데이터 첫인상 보고서

> 본 보고서는 원본 행이 아닌 **구조 점검 요약 정보**를 기준으로 작성했습니다.
> 실제 값의 분포, 날짜 범위, 범주형 고유값 등 요약에 포함되지 않은 내용은 확인되지 않은 상태로 구분했습니다.

## 1. 데이터 파일

| 데이터셋          | 한 행의 의미          |   행 |  열 | 주요 ID                                     |
| ------------- | ---------------- | --: | -: | ----------------------------------------- |
| `customers`   | 고객 1명            | 150 |  6 | `customer_id`                             |
| `products`    | 상품 1개            | 100 |  4 | `product_id`                              |
| `orders`      | 주문 1건            | 301 |  5 | `order_id`, `customer_id`                 |
| `order_items` | 주문에 포함된 상품 항목 1건 | 765 |  5 | `order_item_id`, `order_id`, `product_id` |

### 데이터셋별 해석

* `customers`는 `customer_id`를 기준으로 고객 한 명의 기본 정보를 담는 데이터로 보입니다.
* `products`는 `product_id`를 기준으로 상품 한 개의 이름, 카테고리, 가격을 담는 데이터로 보입니다.
* `orders`는 `order_id`를 기준으로 주문 한 건의 고객, 주문일, 결제수단, 주문상태를 담는 데이터로 보입니다.
* `order_items`는 주문 한 건에 포함된 상품별 수량과 단가를 기록한 주문 상세 데이터로 보입니다.

`orders`와 `order_items`는 데이터의 단위가 다릅니다. 한 주문에 여러 상품이 포함될 수 있으므로 `order_items`에서는 하나의 `order_id`가 여러 행에 반복될 수 있습니다.

---

## 2. 컬럼과 타입

### `customers`

| 컬럼            | 타입      | 해석        |
| ------------- | ------- | --------- |
| `customer_id` | `int64` | 고객 식별자    |
| `name`        | `str`   | 고객 이름     |
| `gender`      | `str`   | 성별 관련 범주값 |
| `age`         | `int64` | 고객 나이     |
| `city`        | `str`   | 고객 도시     |
| `signup_date` | `str`   | 가입일 문자열   |

### `products`

| 컬럼             | 타입      | 해석      |
| -------------- | ------- | ------- |
| `product_id`   | `int64` | 상품 식별자  |
| `product_name` | `str`   | 상품명     |
| `category`     | `str`   | 상품 카테고리 |
| `price`        | `int64` | 상품 가격   |

### `orders`

| 컬럼               | 타입      | 해석               |
| ---------------- | ------- | ---------------- |
| `order_id`       | `int64` | 주문 식별자           |
| `customer_id`    | `int64` | 주문 고객을 연결하는 외래 키 |
| `order_date`     | `str`   | 주문일 문자열          |
| `payment_method` | `str`   | 결제수단             |
| `order_status`   | `str`   | 주문상태             |

### `order_items`

| 컬럼              | 타입      | 해석            |
| --------------- | ------- | ------------- |
| `order_item_id` | `int64` | 주문 상세 행 식별자   |
| `order_id`      | `int64` | 주문을 연결하는 외래 키 |
| `product_id`    | `int64` | 상품을 연결하는 외래 키 |
| `quantity`      | `int64` | 주문 수량         |
| `unit_price`    | `int64` | 주문 상세에 기록된 단가 |

ID 컬럼은 정수형이지만 수치 계산 대상이 아니라 **식별자 또는 연결 키**로 해석해야 합니다. 따라서 ID의 평균, 합계, 크기 순서에는 일반적으로 분석 의미를 부여하지 않습니다.

---

## 3. 결측치

| 데이터셋          | 전체 결측치 수 | 판정     |
| ------------- | -------: | ------ |
| `customers`   |        0 | 결측치 없음 |
| `products`    |        0 | 결측치 없음 |
| `orders`      |        0 | 결측치 없음 |
| `order_items` |        0 | 결측치 없음 |

구조 요약상 네 데이터셋 모두 결측치가 발견되지 않았습니다.

다만 외래 키가 참조 테이블에 존재하지 않는 문제는 일반적인 결측치와는 다른 **참조 무결성 문제**이므로 별도로 확인해야 합니다.

**pandas 확인 목적:**
컬럼별 결측치 수와 비율을 다시 계산하여 구조 점검 요약과 실제 데이터가 일치하는지 검증합니다.

---

## 4. 전체 중복

| 데이터셋          | 완전히 동일한 중복 행 | 판정         |
| ------------- | -----------: | ---------- |
| `customers`   |            0 | 전체 행 중복 없음 |
| `products`    |            0 | 전체 행 중복 없음 |
| `orders`      |            0 | 전체 행 중복 없음 |
| `order_items` |            0 | 전체 행 중복 없음 |

모든 컬럼값이 동일한 완전 중복 행은 발견되지 않았습니다.

다만 특정 컬럼의 값이 반복되는 것과 전체 행 중복은 구분해야 합니다. 다음 반복은 업무 구조상 정상일 수 있습니다.

* 여러 주문을 한 고객이 생성해 `orders.customer_id`가 반복되는 경우
* 한 주문에 여러 상품이 포함되어 `order_items.order_id`가 반복되는 경우
* 동일 상품이 여러 주문에서 판매되어 `order_items.product_id`가 반복되는 경우
* 서로 다른 고객이 동일한 이름을 사용하는 경우

**pandas 확인 목적:**
전체 컬럼 기준 중복과 일부 컬럼 조합 기준 반복을 구분하고, 정상적인 관계 반복을 중복 오류로 잘못 처리하지 않도록 확인합니다.

---

## 5. 주요 ID 결측·중복

| 데이터셋          | 기본 키            | 결측 | 중복 | 고유값 수 | 판정 |
| ------------- | --------------- | -: | -: | ----: | -- |
| `customers`   | `customer_id`   |  0 |  0 |   150 | 정상 |
| `products`    | `product_id`    |  0 |  0 |   100 | 정상 |
| `orders`      | `order_id`      |  0 |  0 |   301 | 정상 |
| `order_items` | `order_item_id` |  0 |  0 |   765 | 정상 |

각 기본 키의 고유값 수가 해당 데이터셋의 행 수와 같습니다. 따라서 현재 요약 기준으로 기본 키는 각 행을 정상적으로 식별하고 있습니다.

### 외래 키

| 외래 키                     | 참조 대상                   | 역할           |
| ------------------------ | ----------------------- | ------------ |
| `orders.customer_id`     | `customers.customer_id` | 주문과 고객 연결    |
| `order_items.order_id`   | `orders.order_id`       | 주문 상세와 주문 연결 |
| `order_items.product_id` | `products.product_id`   | 주문 상세와 상품 연결 |

외래 키는 기본 키와 달리 반복될 수 있습니다. 외래 키의 반복만으로 중복 오류라고 판단해서는 안 됩니다.

**pandas 확인 목적:**
기본 키의 결측과 중복을 검증하고, 외래 키는 고유성보다 참조 대상 존재 여부를 중심으로 점검합니다.

---

## 6. 숫자형 범위

구조 요약에는 숫자형 컬럼의 최솟값, 최댓값, 중앙값, 사분위수 등이 포함되어 있지 않아 실제 범위는 확인할 수 없습니다.

### 범위 점검 대상

| 데이터셋          | 컬럼           | 확인할 내용                              |
| ------------- | ------------ | ----------------------------------- |
| `customers`   | `age`        | 0 이하, 비현실적으로 큰 값, 특정 값에 과도하게 집중된 현상 |
| `products`    | `price`      | 음수 가격, 0원 가격, 지나치게 큰 값              |
| `order_items` | `quantity`   | 0 이하 수량, 지나치게 큰 주문 수량               |
| `order_items` | `unit_price` | 음수 단가, 0원 단가, 지나치게 큰 값              |

`customer_id`, `product_id`, `order_id`, `order_item_id`는 정수형이지만 식별자이므로 숫자형 범위 분석 대상에서 제외하는 것이 적절합니다.

`price`와 `unit_price`의 0도 즉시 오류로 단정해서는 안 됩니다. 실제 업무에서 허용되는 값인지 확인한 후 판단해야 합니다.

**pandas 확인 목적:**
분석 대상 숫자형 컬럼의 최솟값, 최댓값, 분위수와 빈도 분포를 확인하여 비정상 범위와 이상값 후보를 찾습니다.

---

## 7. 범주형 값

구조 요약에는 범주형 컬럼의 실제 고유값과 빈도가 포함되어 있지 않습니다.

### 점검 대상

| 데이터셋        | 컬럼               | 확인할 내용                      |
| ----------- | ---------------- | --------------------------- |
| `customers` | `gender`         | 표기 종류, 공백, 대소문자 또는 표현 방식 혼재 |
| `customers` | `city`           | 도시명 표기 일관성, 앞뒤 공백           |
| `products`  | `category`       | 카테고리 오탈자와 유사 카테고리 중복        |
| `orders`    | `payment_method` | 결제수단 표기 방식의 일관성             |
| `orders`    | `order_status`   | 주문상태 종류와 표기 일관성             |

실제 값이 제공되지 않았으므로 어떤 값이 정상이고 어떤 값이 오류인지는 현재 판단할 수 없습니다.

예를 들어 비슷한 표현이 여러 개 발견되더라도 실제로 서로 다른 상태인지 먼저 업무 의미를 확인해야 합니다.

**pandas 확인 목적:**
각 범주형 컬럼의 고유값, 값별 건수, 앞뒤 공백 여부를 확인하여 오탈자와 표기 불일치 후보를 찾습니다.

---

## 8. 날짜 범위와 변환 실패

날짜 관련 컬럼은 다음과 같습니다.

| 데이터셋        | 컬럼            | 현재 타입 |
| ----------- | ------------- | ----- |
| `customers` | `signup_date` | `str` |
| `orders`    | `order_date`  | `str` |

두 날짜 컬럼 모두 문자열로 저장되어 있습니다. 구조 요약에는 실제 날짜의 시작일, 종료일, 변환 실패 건수가 포함되어 있지 않으므로 날짜 범위는 확인할 수 없습니다.

### 추가 확인 사항

* 날짜형으로 변환되지 않는 값이 있는지
* 여러 날짜 형식이 혼재하는지
* 존재하지 않는 날짜가 있는지
* 날짜 앞뒤에 공백이 있는지
* 가입일과 주문일의 범위가 분석 목적에 적절한지
* 주문일이 고객 가입일보다 앞선 행이 있는지

마지막 항목은 두 데이터셋에 실제로 존재하는 `customer_id`, `signup_date`, `order_date`를 연결해 확인할 수 있습니다. 다만 해당 관계가 반드시 오류를 의미하는지는 데이터 생성 기준을 확인한 후 판단해야 합니다.

**pandas 확인 목적:**
문자열 날짜를 날짜형으로 변환하면서 변환 실패 값을 찾고, 변환에 성공한 값의 최소·최대 날짜와 시간적 순서를 검증합니다.

---

## 9. 파일 간 참조 무결성

| 외래 키 관계                  | 참조 대상                   | 참조되지 않는 값 수 | 판정     |
| ------------------------ | ----------------------- | ----------: | ------ |
| `orders.customer_id`     | `customers.customer_id` |           1 | 불일치 존재 |
| `order_items.order_id`   | `orders.order_id`       |           1 | 불일치 존재 |
| `order_items.product_id` | `products.product_id`   |           1 | 불일치 존재 |

세 외래 키 관계에서 **각각 참조 불일치가 1건씩 확인되었습니다.**

다만 이 수치는 관계별 불일치 건수입니다. 실제 고유 문제 행이 정확히 3개라고 단정할 수는 없습니다.

예를 들어 하나의 `order_items` 행에서 `order_id`와 `product_id`가 동시에 일치하지 않는다면 관계 위반은 2건이지만 실제 문제 행은 1행일 수 있습니다. 따라서 원본 데이터 확인 전에는 다음과 같이 구분해야 합니다.

* 관계별 불일치: 각각 1건
* 실제 고유 문제 행 수: 현재 요약만으로 확인 불가

**pandas 확인 목적:**
각 외래 키를 참조 테이블의 기본 키와 비교하여 불일치 행을 추출하고, 여러 관계의 불일치가 동일한 행에서 발생했는지 확인합니다.

---

## 10. 발견한 문제

### 확인된 문제

1. `orders.customer_id`에서 고객 데이터에 존재하지 않는 참조가 1건 있습니다.
2. `order_items.order_id`에서 주문 데이터에 존재하지 않는 참조가 1건 있습니다.
3. `order_items.product_id`에서 상품 데이터에 존재하지 않는 참조가 1건 있습니다.

### 추가 확인이 필요한 사항

1. `signup_date`와 `order_date`가 문자열로 저장되어 있어 날짜 변환 가능 여부를 확인해야 합니다.
2. 숫자형 분석 컬럼의 실제 범위가 제공되지 않아 음수, 0, 극단값 여부를 확인해야 합니다.
3. 범주형 컬럼의 고유값이 제공되지 않아 오탈자나 표기 혼재 여부를 확인해야 합니다.
4. `products.price`와 `order_items.unit_price`의 차이가 어느 정도 발생하는지 확인해야 합니다.
5. `orders`에는 존재하지만 `order_items`가 연결되지 않는 주문이 있는지 반대 방향 관계도 확인해야 합니다.
6. 주문 상세에서 서로 다른 `order_item_id`가 동일한 주문·상품·수량·단가 조합을 가지는지 확인해야 합니다.

현재 기본 키 결측, 기본 키 중복, 전체 행 중복, 일반 결측치는 발견되지 않았습니다.

---

## 11. 즉시 처리하지 않은 이유

현재 발견된 외래 키 불일치 행은 원인을 확인하지 않고 즉시 삭제해서는 안 됩니다.

가능한 원인으로는 다음과 같은 가설을 검토할 수 있습니다.

* 데이터셋별 추출 시점이 다름
* 기준 데이터 일부가 누락됨
* 과거 기록이 현재 기준 데이터에 남아 있지 않음
* ID 입력 또는 적재 과정에서 오류가 발생함
* 테스트 또는 별도 관리 대상 데이터가 포함됨

이러한 가능성은 현재 확인된 사실이 아니라 원인 조사를 위한 가설입니다.

또한 다음 값들도 자동으로 삭제하거나 수정해서는 안 됩니다.

* 반복되는 외래 키
* 동일한 고객 이름
* 동일한 상품 이름
* 0원 가격 또는 단가
* 동일한 주문·상품 조합
* `products.price`와 다른 `order_items.unit_price`

각 값이 실제 오류인지 정상적인 업무 기록인지 확인할 수 없기 때문입니다. 따라서 우선 삭제하지 않고 검토 대상으로 분리하는 것이 안전합니다.

---

## 12. 다음 전처리 단계에서 결정할 내용

### 1순위: 참조 무결성 처리 기준

* 관계별 불일치 행을 실제로 확인합니다.
* 여러 관계의 문제가 동일한 행에서 발생했는지 확인합니다.
* ID 오타인지, 기준정보 누락인지, 추출 시점 차이인지 분류합니다.
* 수정, 보완, 제외 또는 별도 보관 중 처리 방식을 결정합니다.

### 2순위: 날짜 변환 기준

* `signup_date`와 `order_date`를 날짜형으로 변환합니다.
* 변환 실패 값을 별도로 관리할지 결정합니다.
* 분석 기준 기간과 날짜 범위를 결정합니다.
* 가입일과 주문일의 시간 순서 검증 기준을 정합니다.

### 3순위: 숫자형 유효 범위

* `age`, `price`, `quantity`, `unit_price`의 허용 범위를 정합니다.
* 0과 음수를 오류로 볼지 업무 규칙을 확인합니다.
* 극단값을 제거할지, 유지할지, 별도 표시할지 결정합니다.

### 4순위: 범주형 표준화

* `gender`, `city`, `category`, `payment_method`, `order_status`의 고유값을 확인합니다.
* 앞뒤 공백과 표기 차이를 정리합니다.
* 비슷한 값을 하나로 통합할지 실제 업무 의미를 확인한 후 결정합니다.

### 5순위: 가격 컬럼 사용 기준

* `products.price`와 `order_items.unit_price`의 차이를 확인합니다.
* 상품 기준 가격과 주문 당시 단가 중 어떤 값을 분석에 사용할지 결정합니다.
* 주문 금액 계산 시 `quantity`와 `unit_price`를 사용할지 검토합니다.

### 6순위: 조인과 집계 기준

* 고객–주문은 고객 1명 대 주문 여러 건으로 연결합니다.
* 주문–주문 상세는 주문 1건 대 상세 여러 건으로 연결합니다.
* 상품–주문 상세는 상품 1개 대 상세 여러 건으로 연결합니다.
* 병합 후 행 수가 증가하더라도 이를 주문 수 증가로 잘못 해석하지 않도록 합니다.
* 주문 건수는 필요에 따라 고유 `order_id`를 기준으로 집계합니다.